File: 2_add_two_numbers/Solution.py
# Definition for singly-linked list.
class ListNode:
  def __init__(self, val=0, next=None):
    self.val = val
    self.next = next

class Solution:
  def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    l3 = ListNode()
    n1 = l1
    n2 = l2
    n3 = l3
    
    n3.val = n1.val + n2.val
    while True:
      if n1.next is None and n2.next is None:
        if n3.val > 9:
          n3.next = ListNode(val=1)
          n3.val -= 10
        return l3
      elif n1.next is None:
        n2 = n2.next
        n3.next = ListNode(val=n2.val)
      elif n2.next is None:
        n1 = n1.next
        n3.next = ListNode(val=n1.val)
      else:
        n1 = n1.next
        n2 = n2.next
        n3.next = ListNode(val=n1.val + n2.val)

      if n3.val > 9:
        n3.next.val += 1
        n3.val -= 10
        
      n3 = n3.next
        

File: 2_add_two_numbers/Solution.cpp

//Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    ListNode* l3 = new ListNode();
    ListNode* n1 = l1;
    ListNode* n2 = l2;
    ListNode* n3 = l3;
    int carry = 0;
    
    while (n1 || n2) {
      int x = n1 ? n1->val : 0;
      int y = n2 ? n2->val : 0;

      n3->val = (x + y + carry) % 10;
      carry = (x + y + carry) > 9 ? 1 : 0;

      n1 = n1 ? n1->next : n1;
      n2 = n2 ? n2->next : n2;

      if (n1 || n2) {
        n3->next = new ListNode();
        n3 = n3->next;
      } else if (carry) {
        n3->next = new ListNode(carry);
      }
    }

    return l3;
  }
};